diff --git a/webrtc/modules/audio_device/ios/audio_device_ios.mm b/webrtc/modules/audio_device/ios/audio_device_ios.mm
index 47503a9..070afff 100644
--- a/webrtc/modules/audio_device/ios/audio_device_ios.mm
+++ b/webrtc/modules/audio_device/ios/audio_device_ios.mm
@@ -1193,8 +1193,8 @@ int32_t AudioDeviceIOS::InitPlayOrRecord() {
     // Non-obviously, setting them to the value they already have will clear
     // transient properties (such as PortOverride) that some other component may
     // have set up.
-    if (session.mode != AVAudioSessionModeVoiceChat) {
-      [session setMode:AVAudioSessionModeVoiceChat error:&error];
+    if (session.mode != AVAudioSessionModeVideoChat) {
+      [session setMode:AVAudioSessionModeVideoChat error:&error];
       if (error != nil) {
         const char* errorString = [[error localizedDescription] UTF8String];
         WEBRTC_TRACE(kTraceInfo, kTraceAudioDevice, _id,
